include/fmt/format-inl.h

```diff
@@ -933,90 +923,22 @@ inline int floor_log10_pow2_minus_log10_4_over_3(int e) FMT_NOEXCEPT {
-      {0xc28f5c29, 0x0a3d70a3}, {0x26e978d5, 0x020c49ba},
-      {0x3afb7e91, 0x0068db8b}, {0x0bcbe61d, 0x0014f8b5},
-      {0x68c26139, 0x000431bd}, {0xae8d46a5, 0x0000d6bf},
-      {0x22e90e21, 0x00002af3}, {0x3a2e9c6d, 0x00000897},
-      {0x3ed61f49, 0x000001b7}};
-  return x * divtest_table[exp].mod_inv <= divtest_table[exp].max_quotient;
-}
-inline bool divisible_by_power_of_5(uint64_t x, int exp) FMT_NOEXCEPT {
-  FMT_ASSERT(exp <= 23, "too large exponent");
-  static constexpr const divtest_table_entry<uint64_t> divtest_table[] = {
-      {0x0000000000000001, 0xffffffffffffffff},
-      {0xcccccccccccccccd, 0x3333333333333333},
-      {0x8f5c28f5c28f5c29, 0x0a3d70a3d70a3d70},
-      {0x1cac083126e978d5, 0x020c49ba5e353f7c},
-      {0xd288ce703afb7e91, 0x0068db8bac710cb2},
-      {0x5d4e8fb00bcbe61d, 0x0014f8b588e368f0},
-      {0x790fb65668c26139, 0x000431bde82d7b63},
-      {0xe5032477ae8d46a5, 0x0000d6bf94d5e57a},
-      {0xc767074b22e90e21, 0x00002af31dc46118},
-      {0x8e47ce423a2e9c6d, 0x0000089705f4136b},
-      {0x4fa7f60d3ed61f49, 0x000001b7cdfd9d7b},
-      {0x0fee64690c913975, 0x00000057f5ff85e5},
-      {0x3662e0e1cf503eb1, 0x000000119799812d},
-      {0xa47a2cf9f6433fbd, 0x0000000384b84d09},
-      {0x54186f653140a659, 0x00000000b424dc35},
-      {0x7738164770402145, 0x0000000024075f3d},
-      {0xe4a4d1417cd9a041, 0x000000000734aca5},
-      {0xc75429d9e5c5200d, 0x000000000170ef54},
-      {0xc1773b91fac10669, 0x000000000049c977},
-      {0x26b172506559ce15, 0x00000000000ec1e4},
-      {0xd489e3a9addec2d1, 0x000000000002f394},
-      {0x90e860bb892c8d5d, 0x000000000000971d},
-      {0x502e79bf1b6f4f79, 0x0000000000001e39},
-      {0xdcd618596be30fe5, 0x000000000000060b}};
-  return x * divtest_table[exp].mod_inv <= divtest_table[exp].max_quotient;
-}
-
-// Replaces n by floor(n / pow(5, N)) returning true if and only if n is
-// divisible by pow(5, N).
-// Precondition: n <= 2 * pow(5, N + 1).
+// Replaces n by floor(n / pow(10, N)) returning true if and only if n is
+// divisible by pow(10, N).
+// Precondition: n <= pow(10, N + 1).
 template <int N>
-bool check_divisibility_and_divide_by_pow5(uint32_t& n) FMT_NOEXCEPT {
+bool check_divisibility_and_divide_by_pow10(uint32_t& n) FMT_NOEXCEPT {
   static constexpr struct {
     uint32_t magic_number;
-    int bits_for_comparison;
-    uint32_t threshold;
-    int shift_amount;
-  } infos[] = {{0xcccd, 16, 0x3333, 18}, {0xa429, 8, 0x0a, 20}};
+    int margin_bits;
+    int divisibility_check_bits;
+  } infos[] = {{0x199a, 8, 8}, {0xa3d71, 10, 16}};
   constexpr auto info = infos[N - 1];
   n *= info.magic_number;
-  const uint32_t comparison_mask = (1u << info.bits_for_comparison) - 1;
-  bool result = (n & comparison_mask) <= info.threshold;
-  n >>= info.shift_amount;
+  n >>= info.margin_bits;
+  const uint32_t comparison_mask = (1u << info.divisibility_check_bits) - 1;
+  bool result = (n & comparison_mask) == 0;
+  n >>= info.divisibility_check_bits;
```

Comment:
Not yet, I'm still working on rewriting the paper. But it's probably simple enough so that I can explain it here. Roughly speaking, those magic numbers are still binary digits of 1/5 and 1/25. (Note that ceil(0xcccd/8) = 0x199a.)

What we are trying to do here is to find integers `m` and `k` and `l` such that

1.  `floor(n/d) = floor(n * m / 2^(k+l))` where `d=10` or `d=100`,
2.  `floor(nm/2^k) mod 2^l = 0` if and only if `n` is divisible by `d`.

(Here, `m` is `magic_number`, `k` is `margin_bits`, and `l` is `divisibility_check_bits`.)

So if these are possible, then we first multiply `n` with the magic number `m` and shift to right by `k`. The lower `l` bits of the resulting number is zero if and only if `n` is divisible by `d`, and then we shift to right again by `l` to wipe out those bits, then the result must be precisely `floor(n/d)`.

The first item is nothing but what the classical paper by Granlund-Montgomery tells us, so let's take it as granted; we set `m` to be the ceiling of `2^(k+l)/d` for large enough `k+l`. (To be a bit more specific, `k+l >= 16` for `float` and `k+l >= 26` for `double` are more than enough.)

The rationale behind the second item is that, since `m` is almost equal to `2^(k+l)/d`, so if `n` is divisible by `d`, then `n * m` should be almost a multiple of `2^(k+l)`. And `2^k` is how far `n * m` can deviate from truly being a multiple of `2^(k+l)`.

To make this idea precise, note that 1 is equivalent to
````
  floor(n/d) <= nm/2^(k+l) < floor(n/d) + 1.
````
Now, let `n = dq + r` with `0 <= r < d`, then this becomes
````
  q <= nm/2^(k+l) < q + 1,
````
so multiplying `2^l` yields
````
  q * 2^l <= nm/2^k < (q+1) * 2^l.
````
Hence, `floor(nm/2^k) mod 2^l` is just `floor(nm/2^k) - q * 2^l`.

Now, we want two things:

1. `floor(nm/2^k) - q * 2^l = 0` if `r = 0`, and
2. `floor(nm/2^k) - q * 2^l != 0` if `r != 0`.

The first one can be rewritten as
````
  q * 2^l <= nm/2^k < q * 2^l + 1,
````
and since we already have the first inequality, it is equivalent to
````
  nm < q * 2^(k+l) + 2^k.
````
Using `n = dq`, we can rewrite it as
````
  q(dm - 2^(k+l)) < 2^k,
````
and since `m = ceil(2^(k+l)/d)`, we have `0 <= dm - 2^(k+l) < d`, so it is sufficient to have
````
  qd = n < 2^k.
````
Since we know the range of `n` (`0~100` for `float` and `0~1000` for `double`), we can easily figure out that `k >= 7` for `float` and `k >= 10` for double are sufficient to ensure that `r = 0` implies `floor(nm/2^k) mod 2^l = 0`.

On the other hand, the second one can be rewritten as
````
  nm/2^k >= q * 2^l + 1,
````
or equivalently,
````
  (dq + r) * m >= q * 2^(k+l) + 2^k,
````
or equivalently,
````
  rm >= q(2^(k+l) - dm) + 2^k.
````
Since `dm - 2^(k+l)  > 0`, we can see that having
````
  m >= 2^k
````
is sufficient to ensure that `r != 0` implies `floor(nm/2^k) mod 2^l != 0`. Or to be a bit more lenient, it is sufficient to have
````
  2^(k+l)/d >= 2^k,
````
or equivalently
````
  2^l >= d.
````
Hence, `l >= 4` for `float` and `l >= 7` for `double` are enough. (Of course, we set `l = 8` for `float` and `l = 16` for `double` to make `mod 2^l` a trivial operation.)

I also did an exhaustive test of this [here](https://github.com/jk-jeon/dragonbox/blob/master/subproject/test/source/verify_magic_division.cpp) to make sure things are done correctly.

Reply:
Thanks a lot for the detailed explanation!

Reply:
My pleasure! I'll get back to you when I finish writing things up. The paper will contain several sharper results generalizing/unifying this and several other tricks I have been relying on.

Also, if time permits, I will work on an alternative idea about fixed-precision printing, based on a generalization of Dr. Lemire's paper I linked. I hope it ends up with something I can contribute again.

Reply:
So in [this paper](https://r-libre.teluq.ca/2438/1/2012.12369.pdf), a better method is introduced which allows us to save yet one more instruction:

```
  const uint32_t magic_number = (1u << 16) / divisor + 1;
  const uint32_t comparison_mask = (1u << 16) - 1;
  n *= magic_number;
  bool result = (n & comparison_mask) < magic_number;
  n >>= 16;
```
(This method, the one I explained, and the classical divisibility test by Granlund-Montgomery using modular inverse, are in fact all just special cases of a more general result that will appear in the paper I'm writing. Basically, what this method is doing is not very different from what I explained. Indeed, `floor(nm/2^k) mod 2^l = 0` is equivalent to `nm mod 2^(k+l) < 2^k`, and the thing is that there is no terribly important reason to have a power of 2 on the right-hand side now. And it turns out that we can replace it by `m`.)

I don't think this code will perform measurably better, but it's shorter and presumably better, so why not.